File: emulinker/src/main/java/org/emulinker/net/PrivateUDPServer.java
package org.emulinker.net;

import static com.codahale.metrics.MetricRegistry.name;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Timer;
import com.google.common.flogger.FluentLogger;
import java.net.*;
import java.nio.ByteBuffer;
import org.emulinker.util.*;

public abstract class PrivateUDPServer extends UDPServer {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private final InetAddress remoteAddress;
  private final Timer clientRequestTimer;

  private InetSocketAddress remoteSocketAddress;

  public PrivateUDPServer(
      boolean shutdownOnExit, InetAddress remoteAddress, MetricRegistry metrics) {
    super(shutdownOnExit, metrics);
    this.remoteAddress = remoteAddress;
    this.clientRequestTimer = metrics.timer(name(this.getClass(), "clientRequests"));
  }

  public InetAddress getRemoteInetAddress() {
    return remoteAddress;
  }

  public InetSocketAddress getRemoteSocketAddress() {
    return remoteSocketAddress;
  }

  @Override
  protected void handleReceived(ByteBuffer buffer, InetSocketAddress inboundSocketAddress) {
    if (remoteSocketAddress == null) remoteSocketAddress = inboundSocketAddress;
    else if (!inboundSocketAddress.equals(remoteSocketAddress)) {
      logger.atWarning().log(
          "Rejecting packet received from wrong address: "
              + EmuUtil.formatSocketAddress(inboundSocketAddress)
              + " != "
              + EmuUtil.formatSocketAddress(remoteSocketAddress));
      return;
    }

    try (final Timer.Context context = clientRequestTimer.time()) {
      handleReceived(buffer);
    }
  }

  protected abstract void handleReceived(ByteBuffer buffer);

  protected void send(ByteBuffer buffer) {
    super.send(buffer, remoteSocketAddress);
  }
}

File: emulinker/src/main/java/org/emulinker/net/PrivateUDPServer.kt
package org.emulinker.net

import com.codahale.metrics.MetricRegistry
import com.codahale.metrics.Timer
import com.google.common.flogger.FluentLogger
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.withTimeout
import org.emulinker.config.RuntimeFlags
import org.emulinker.util.EmuUtil.formatSocketAddress

private val logger = FluentLogger.forEnclosingClass()

abstract class PrivateUDPServer(
    val remoteInetAddress: InetAddress, metrics: MetricRegistry, private val flags: RuntimeFlags
) : UDPServer(flags) {

  private val clientRequestTimer: Timer

  lateinit var remoteSocketAddress: InetSocketAddress
    private set

  override suspend fun handleReceived(
      buffer: ByteBuffer, remoteSocketAddress: InetSocketAddress, requestScope: CoroutineScope
  ) {
    if (!this::remoteSocketAddress.isInitialized) {
      this.remoteSocketAddress = remoteSocketAddress
    } else if (remoteSocketAddress != this.remoteSocketAddress) {
      logger
          .atWarning()
          .log(
              "Rejecting packet received from wrong address. Expected=%s but was %s",
              formatSocketAddress(this.remoteSocketAddress),
              formatSocketAddress(remoteSocketAddress))

      return
    }
    clientRequestTimer.time().use {
      try {
        withTimeout(flags.requestTimeout) { handleReceived(buffer) }
      } catch (e: TimeoutCancellationException) {
        logger.atSevere().withCause(e).log("Request timed out")
      }
    }
  }

  protected abstract suspend fun handleReceived(buffer: ByteBuffer)

  protected suspend fun send(buffer: ByteBuffer) {
    super.send(buffer, remoteSocketAddress)
  }

  init {
    clientRequestTimer = metrics.timer(MetricRegistry.name(this.javaClass, "clientRequests"))
  }
}
